src/pipelinedefinitionstatus.hpp

```diff
@@ -0,0 +1,288 @@
+        return {};
+    }
+    StateKeeper handle(const UsedModelChangedEvent& e) const {
+        SPDLOG_ERROR("State:{} handling: {}. {}", pipelineDefinitionStateCodeToString(getState()), e.name, INVALID_TRANSITION_MESSAGE);
+        throw std::logic_error(INVALID_TRANSITION_MESSAGE);
+        return {};
+    }
+    StateKeeper handle(const RetireEvent& e) const {
+        SPDLOG_ERROR("State:{} handling: {}. {}", pipelineDefinitionStateCodeToString(getState()), e.name, INVALID_TRANSITION_MESSAGE);
+        throw std::logic_error(INVALID_TRANSITION_MESSAGE);
+        return {};
+    }
+};
+
+struct AvailableState {
+    static const PipelineDefinitionStateCode code = PipelineDefinitionStateCode::AVAILABLE;
+    PipelineDefinitionStateCode getState() const {
+        return code;
+    }
+    void print() const {
+        SPDLOG_ERROR(pipelineDefinitionStateCodeToString(getState()));
+    }
+    StateKeeper handle(const ValidationPassedEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+    StateChanger<LoadingPreconditionFailedState> handle(const ValidationFailedEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+    StateChanger<AvailableRequiredRevalidation> handle(const UsedModelChangedEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+    StateChanger<RetiredState> handle(const RetireEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+};
+
+struct AvailableRequiredRevalidation {
+    static const PipelineDefinitionStateCode code = PipelineDefinitionStateCode::AVAILABLE_REQUIRED_REVALIDATION;
+    PipelineDefinitionStateCode getState() const {
+        return code;
+    }
+    void print() const {
+        SPDLOG_ERROR(pipelineDefinitionStateCodeToString(getState()));
+    }
+    StateChanger<AvailableState> handle(const ValidationPassedEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+    StateChanger<LoadingPreconditionFailedState> handle(const ValidationFailedEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+    StateKeeper handle(const UsedModelChangedEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+    StateChanger<RetiredState> handle(const RetireEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+};
+
+struct LoadingPreconditionFailedState {
```

Comment:
What about adding comments to each state with its meaning at runtime?

Reply:
For example this state indicates that validation occurred and failed due to problems with underlying model/models.

Reply:
That's good point. I want tough to first handle the logic of runtime modification & then work with error logging. The error message & explanation is part of getModelStatus response so I think we may either move this to that task or keep it as a separate MR for this ticket.

Reply:
ok I will add comments in code